fpga_compile code sample (#1)
KevinUTAT/oneAPI-samples: DirectProgramming/C++SYCL_FPGA/Tutorials/GettingStarted/fpga_compile/README.md

Long compile times are detrimental to developer productivity. The Intel® oneAPI DPC++/C++ Compiler provides several mechanisms that enable developers targeting FPGAs to iterate quickly on their designs. By circumventing the time-consuming process of full FPGA compilation wherever possible, developers can enjoy the fast compile times familiar to CPU and GPU developers.

### Full-system Acceleration vs Kernels Reusable IP Component

In the FPGA full-system acceleration flow, you can generate an executable host application and acceleration kernel for a PCIe FPGA board if you have a compatible board support package (BSP). Intel provides BSPs for the Intel® PAC with Intel Arria® 10 GX FPGA, and the Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX). If you have a different board, check with your vendor to see if they supply a BSP.

> In the FPGA IP component authoring flow, you can generate a reusable IP component that you can import into an Intel® Quartus® Prime project. You can generate an IP by targeting your compilation to a supported Intel® FPGA device family or part number (for example, `Agilex` or `AGFB014R24A3EV`) instead of a named board (for example, `intel_a10gx_pac:pac_a10`). 

Author: Interesting. I am getting 
```
Error: 'AGFB014R24A3EV' does not specify a valid device, device family, or board variant.
Please make sure the option -Xstarget points to a valid board variant
```

> In the FPGA IP component authoring flow, you can generate a reusable IP component that you can import into an Intel® Quartus® Prime project. You can generate an IP by targeting your compilation to a supported Intel® FPGA device family or part number (for example, `Agilex` or `AGFB014R24A3EV`) instead of a named board (for example, `intel_a10gx_pac:pac_a10`). 

Author: i am changing the example to `AGFA014R24B1E1V`, this number seems to work with 2023.1 rel/

> The FPGA IP component authoring flow does not generate any FPGA accelerated executable, only RTL (Register Transfer Level) IP component files. The host application is  treated only as a 'testbench' that exercises and validates your IP component in emulation and simulation.

Author: Full-system design => Multiarchitecture Binary
IP Authoring => IP component generation

This compilation process takes hours, although it may be faster if you generate a re-usable IP component.

### Device Selectors
The following code snippet demonstrates how you can specify the target device in your source code. The selector is used to specify the target device at runtime.

It is recommended to use a preprocessor define to choose between the emulator and FPGA selectors. This makes it easy to switch between targets using only command-line options. Since the FPGA only supports ahead-of-time compilation, dynamic selectors (such as the default_selector) are less useful than explicit selectors when targeting FPGA.

```c++
// FPGA device selectors are defined in this utility header
#include <sycl/ext/intel/fpga_extensions.hpp>

int main() {

// choose a selector based on compiler flags.
#if defined(FPGA_EMULATOR)
        sycl::queue q(sycl::ext::intel::fpga_emulator_selector_v);
#elif defined(FPGA_SIMULATOR)
        sycl::queue q(sycl::ext::intel::fpga_simulator_selector_v);
#else // FPGA_HARDWARE
        sycl::queue q(sycl::ext::intel::fpga_selector_v);
#endif

>     ...

Reviewer: This does not reflect what's in the code

| `-Xshardware`                    | Instructs the compiler to target FPGA hardware and generate an FPGA bitstream. Cannot be used along with `-Xssimulator`.
| `-Xssimulator`                   | Instructs the compiler to generate a co-simulation executable instead of an FPGA bitstream. Cannot be used along with `-Xshardware`.
| `-fsycl-link=early`              | Instructs the compiler to stop after creating the FPGA early image (and associated optimization report). No executable (co-simulation or FPGA bitstream) is created. Must use with `-Xshardware`
| <span id="families">`-Xstarget=<FPGA device family>`</span> | Instructs the compiler to target FPGA device family. Valid values are `Agilex`, `Arria10`, and `Stratix10`.<br> If you omit the `-Xstarget` flag, the compiler chooses the default FPGA board variant `pac_a10 from the intel_a10gx_pac` BSP (equivalent to `-Xstarget=intel_a10gx_pac:pac_a10`).
| <span id="partnumbers">`-Xstarget=<FPGA part number>`</span>   | Instructs the compiler to target FPGA part number (sometimes called an OPN). You can specify any valid Intel® Agilex™, Arria® 10, or Intel® Stratix® 10 part number.

> | <span id="bsps">`-Xstarget=<bsp:variant>`</span>  | Instructs the compiler to target an FPGA board variant and BSP.

Reviewer 2: Can we link to the documentation instead of reproducing all the flags here? This seems a bit extra.

### Source Code

> There are 3 parts to this tutorial located in the 3 sub-folders. Together, they demonstrate how an algorithm can migrate from vanilla C++ code to SYCL for FPGA.

Reviewer 2: ```suggestion
There are 3 parts to this tutorial located in the 3 sub-folders. Together, they demonstrate how you can migrate an algorithm from vanilla C++ code to SYCL for FPGA. Note that you may also choose to use a functor with buffers, or a function with USM.
```

#### Part 1 C++
A vector addition program in vanilla C++. Can be compiled using `icpx vector_add.cpp`.

#### Part 2 SYCL* (functor and USM)

> The same vector addition in SYCL* C++  with a 'functor' coding style using a unified shared memory (USM) interface. Compare with the source code in part 1 to see SYCL*-specific code changes. This code style is recommended for IP component designs because it clearly separates device code from host/testbench code.

Reviewer 2: ```suggestion
The same vector addition in SYCL* C++  with a 'functor' coding style using a unified shared memory (USM) interface. Compare with the source code in part 1 to see SYCL*-specific code changes. Observe how the `VectorAdd` functor is called using `q.single_task<...>(VectorAdd{...});`. This tells the DPC++ compiler to convert `VectorAdd` into RTL. Also observe how `vec_a`, `vec_b`, and `vec_c` are allocated into a shared memory space using `malloc_shared`; this tells the DPC++ compiler that `vec_a`, `vec_b`, and `vec_c` should be visible both to your kernel and your host code (or if you are creating an IP component, the testbench code).
```

This coding style does not really improve the clarity of device/host code separation.

  >  ```
  >  cmake -G "NMake Makefiles" .. -DFPGA_DEVICE=<board-support-package>:<board-variant>
  >  ``` 
  >
  > You will only be able to run an executable on the FPGA if you specified a BSP.

### Build using `make`/`nmake`
After using CMake to generate build artifacts, you can then build with specific targets. This project can build 4 targets.

| Compilation Type    | Command (Linux)    | Command (Windows)
|:---                 |:---                |:---
| FPGA Emulator       | `make fpga_emu`    | `nmake fpga_emu`
| Optimization Report | `make report`      | `nmake report`
| FPGA Simulator      | `make fpga_sim`    | `nmake fpga_sim`
| FPGA Hardware       | `make fpga`        | `nmake fpga`

The `fpga_emu`, `fpga_sim` and `fpga` targets produce binaries that you can run. The executables will be called `vector_add.fpga_emu`, `vector_add.fpga_sim`, and `vector_add.fpga`. The `fpga` target will produce an executable binary if you create a full-system acceleration kernel.

For part 1 of this tutorial, only the `fpga_emu` target is available as this regular C++ code is only capable of targeting a CPU.

## Examining the Reports

> In *part2-dcpcpp_functor_usm* and *part3-dcpcpp_lambda_buffers*, after running the `report` target, locate `report.html` in the `vector_add.report.prj/reports/` directory. Open the report in any of Chrome*, Firefox*, Edge*, or Internet Explorer*.

Author: I will change to using `fpga_report`

Long compile times are detrimental to developer productivity. The Intel® oneAPI DPC++/C++ Compiler provides several mechanisms that enable developers targeting FPGAs to iterate quickly on their designs. By circumventing the time-consuming process of full FPGA compilation wherever possible, developers can enjoy the fast compile times familiar to CPU and GPU developers.

> ### Multiarchitecture binary vs reusable IP components

Reviewer 2: ```suggestion
### Multiarchitecture binary vs IP component
```

> In the FPGA multiarchitecture binary generation flow, you can generate an executable host application and acceleration kernel for a PCIe FPGA board if you have a compatible board support package (BSP). Intel provides BSPs for the Intel® PAC with Intel Arria® 10 GX FPGA, and the Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX). If you have a different board, check with your vendor to see if they supply a BSP.

Reviewer 2: ```suggestion
In the FPGA multiarchitecture binary generation flow, you can generate an executable host application and accelerator for a PCIe FPGA board if you have a compatible board support package (BSP). Intel provides BSPs for the Intel® PAC with Intel Arria® 10 GX FPGA, and the Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX). If you have a different board, check with your vendor to see if they supply a BSP.
```

> In the FPGA IP component generation flow, you can generate a reusable IP component that you can import into an Intel® Quartus® Prime project. You can generate an IP by targeting your compilation to a supported Intel® FPGA device family or part number (for example, `Agilex` or `AGFA014R24B1E1V`) instead of a named board (for example, `intel_a10gx_pac:pac_a10`). 

Reviewer 2: ```suggestion
In the FPGA IP component generation flow, you can generate an IP component that you can import into an Intel® Quartus® Prime project. You can generate an IP by targeting your compilation to a supported Intel® FPGA device family or part number (for example, `Agilex` or `AGFA014R24B1E1V`) instead of a named board (for example, `intel_a10gx_pac:pac_a10`). 
```

> The FPGA IP component generation flow does not generate any FPGA accelerated executable, only RTL (Register Transfer Level) IP component files. The host application is  treated only as a 'testbench' that exercises and validates your IP component in emulation and simulation.

Reviewer 2: ```suggestion
The FPGA IP component generation flow does not generate any FPGA accelerated executable, only RTL (Register Transfer Level) IP component files. The host application is treated only as a 'testbench' that exercises and validates your IP component in emulation and simulation.
```

   | Simulator | Minutes        | RTL + FPGA reports + x86-64 binary                                           | Compiles the FPGA device code to RTL and generates a simulation testbench. Use the Questa*-Intel® FPGA Edition simulator to verify your design.
   | FPGA Hardware | Multiple Hours | Quartus Place & Route (Multiarchitecture binary) + FPGA reports + x86-64 host binary | Compiles the FPGA device code to RTL and compiles the generated RTL using Intel® Quartus® Prime. If you specified a BSP with `FPGA_DEVICE`, this will generate an FPGA image that you can run on the corresponding accelerator board.

The typical FPGA development workflow is to iterate in each of these stages, refining the code using the feedback provided by that stage. You can avoid long compile times by relying on emulation and the optimization report whenever possible.

#### FPGA Emulator

> The FPGA emulator is the fastest method to verify the correctness of your code. The FPGA emulator executes the SYCL* device code on the CPU. The emulator is similar to the SYCL* host device, but unlike the host device, the FPGA emulator device supports FPGA extensions such as FPGA pipes and `fpga_reg` (although some of these feature, such as `fpga_reg` may not affect how your design runs on the emulator).

Reviewer 2: ```suggestion
The FPGA emulator is the fastest method to verify the correctness of your code. The FPGA emulator executes the SYCL* device code on the CPU. The emulator is similar to the SYCL* host device, but unlike the host device, the FPGA emulator device supports FPGA extensions such as FPGA pipes and `fpga_reg` (although some of these features, such as `fpga_reg` may not affect how your design runs on the emulator).
```

* **Undefined behavior may differ.** If your code produces different results when compiled for the FPGA emulator versus FPGA hardware, your code may exercises undefined behavior. By definition, undefined behavior is not specified by the language specification and may manifest differently on different targets.

#### Optimization Report (Early Image)

For this compilation type, your SYCL device code is optimized and converted into an FPGA design specified in Verilog RTL (a low-level, native entry language for FPGAs). This intermediate compilation result is also called the *FPGA early device image*, which is **not** executable. 

The optimization report contains significant information about how the compiler has transformed your device code into an FPGA design. The report includes visualizations of structures generated on the FPGA, performance and expected performance bottleneck information, and estimated resource utilization. Optimization reports are generated for the "optimization report", "simulator" and "hardware" compilation types. 

> The [FPGA Optimization Guide for Intel® oneAPI Toolkits Developer Guide][1] contains a chapter about how to analyze the reports generated after the FPGA early image and FPGA image.

Reviewer 2: fancy markdown syntax 👍👍👍

#### FPGA Simulator
The FPGA simulator allows you to simulate the exact behavior of the synthesized kernel. Like emulation, you can run simulation on a system that does not have a target FPGA board installed. The simulator models a kernel much more accurately than the emulator, but it is much slower than the emulator.

The Intel oneAPI DPC++/C++ Compiler links your design C++ testbench with an RTL-compiled version of your component that runs in an RTL simulator. You do not need to invoke any RTL simulator manually, but you can add the `-Xsghdl` flag to save the simulation waveform for later viewing.

> **Note**: Running the simulation executable can take a long time if your device code is complex or if your test inputs are large. To save simulation time, use the smallest possible input.

#### FPGA Hardware (Hardware Image)

The generated Verilog RTL is mapped onto the FPGA hardware resources by the Intel® Quartus® Prime software. The estimated performance and resource utilization is therefore much more accurate than the estimates obtained in the optimization report compilation type. 

> If the compilation flow is for the multiarchitecture binary, the result is an FPGA hardware binary (also referred to as a bitstream) that is executable if you target a supported BSP. The compiler will interface your design with the BSP, your host application will seamlessly make the system calls to launch kernels on the FPGA. 

Reviewer 2: ```suggestion
If you compile a multiarchitecture binary, the resulting binary will include an FPGA hardware image (also referred to as a bitstream) that is executable on an FPGA accelerator card with a supported BSP. The compiler will interface your design with the BSP, and your host application will seamlessly make the system calls to launch kernels on the FPGA. 
```

> If the compilation flow is for reusable IP component, the compilation result is **not** executable. IP components are compiled in isolation and not interfaced with other components on the FPGA. The purpose of this compilation flow is to get accurate resource utilization and performance data for IP components.

Reviewer 2: ```suggestion
If you compile an IP component, the compilation result is **not** executable. IP components are compiled in isolation and not interfaced with other components on the FPGA. The purpose of this compilation flow is to get accurate resource utilization and performance data for IP components.
```

### Device Selectors
The following code snippet demonstrates how you can specify the target device in your source code. The selector is used to specify the target device at runtime.

> It is recommended to use a preprocessor define to choose between the emulator and FPGA selectors. This makes it easy to switch between targets using only command-line options. Since the FPGA only supports ahead-of-time compilation, dynamic selectors (such as the `default_selector`) are less useful than explicit selectors when targeting FPGA.

Reviewer 2: ```suggestion
It is recommended to use a preprocessor macro to choose between the emulator and FPGA selectors. This makes it easy to switch between targets using only command-line options. Since the FPGA only supports ahead-of-time compilation, dynamic selectors (such as the `default_selector`) are less useful than explicit selectors when targeting FPGA.
```

### Compiler Options
This section includes a helpful list of commands and options to compile this design for the FPGA emulator, generate the FPGA early image optimization reports, and compile for FPGA hardware.\

> >**Note:** In this sample, the compiler is refered to as `icpx`. On Windows, you should use `icx-cl`.

Reviewer 2: ```suggestion
> **Note**: In this sample, the compiler is refered to as `icpx`. On Windows, you should use `icx-cl`.
```

```
Through `-Xstarget`, you can target an explicit board, a device family or a FPGA part number.

Hardware

```bash
# FPGA hardware image:
icpx -fsycl -fintelfpga -DFPGA_HARDWARE fpga_design.cpp -Xshardware -Xstarget=Agilex -o fpga_design.fpga
```
Through `-Xstarget`, you can target an explicit board, a device family or a FPGA part number.

`-DFPGA_EMULATOR`, `-DFPGA_SIMULATOR`, `-DFPGA_HARDWARE` are options that adds a preprocessor define that invokes the emulator/simulator/FPGA device selector in this sample (see code snippet above).

The [Intel® oneAPI Programming Guide][2] contains a chapter explains the compiler options used here.

### Source Code

There are 5 parts to this tutorial located in the 3 sub-folders. Together, they demonstrate how you can migrate an algorithm from vanilla C++ code to SYCL for FPGA. Note that you may also choose to use a functor with buffers, or a function with USM.

#### Part 1 C++
Part 1 demonstrates a vector addition program in vanilla C++. Observe how the `VectorAdd` function is separated from the `main()` function, and the `vec_a`, `vec_b`, and `vec_c` vectors are allocated onto the heap.

#### Part 2 SYCL* (functor and USM)

> The same vector addition in SYCL* C++  with a 'functor' coding style using unified shared memory (USM) interface. Compare with the source code in part 1 to see SYCL*-specific code changes. Observe how the `VectorAdd` functor is called using `q.single_task<...>(VectorAdd{...});`. This tells the DPC++ compiler to convert `VectorAdd` into RTL. Also observe how `vec_a`, `vec_b`, and `vec_c` are allocated into a shared memory space using `malloc_shared`; this tells the DPC++ compiler that `vec_a`, `vec_b`, and `vec_c` should be visible both to your kernel and your host code (or if you are creating an IP component, the testbench code).

Reviewer 2: ```suggestion
Part 2 shows the same vector addition from part 1, but in SYCL* C++  with a 'functor' coding style using a unified shared memory (USM) interface. Compare with the source code in part 1 to see SYCL*-specific code changes. Observe how the `VectorAdd` functor is called using `q.single_task<...>(VectorAdd{...});`. This tells the DPC++ compiler to convert `VectorAdd` into RTL. Also observe how `vec_a`, `vec_b`, and `vec_c` are allocated into a shared memory space using `malloc_shared`; this tells the DPC++ compiler that `vec_a`, `vec_b`, and `vec_c` should be visible both to your kernel and your host code (or if you are creating an IP component, the testbench code).
```

> #### Part 3 SYCL* (lambda function and buffer)
> The same vector addition in SYCL* C++ with a 'function' coding style and buffer & accessor interface. This code style will be familiar to users who are already experienced with SYCL*. Observe how the `VectorAdd` function is called using a lambda:
> ```c++
> h.single_task<...>([=]() {
>     VectorAdd(...);
> });
> ```
> This tells the DPC++ compiler to convert `VectorAdd` into RTL. Also observe how `vec_a`, `vec_b`, and `vec_c` are copied into buffers before the `VectorAdd` function is called.

Reviewer 2: ```suggestion
#### Part 3 SYCL* (lambda function and buffer)
Part 3 shows the vector addition in SYCL* C++ with a 'function' coding style and buffer & accessor interface. This code style will be familiar to users who are already experienced with SYCL*. Observe how `vec_a`, `vec_b`, and `vec_c` are copied into buffers before the `VectorAdd` function is called.
```

this should be part 4

#### Part 4 SYCL* (functor and buffer)

> The same vector addition in SYCL* C++  with a 'functor' coding style using buffer & accessor interface

Reviewer 2: ```suggestion
Part 4 demonstrates the vector addition in SYCL* C++  with a 'functor' coding style using buffer & accessor interface. Observe how template parameters are used in the functor to pass in the accessors.
```
this should be part 5

> #### Part 5 SYCL* (lambda function and USM)
> The same vector addition in SYCL* C++ with a 'function' coding style using unified shared memory (USM)

Reviewer 2: ```suggestion
#### Part 5 SYCL* (lambda function and USM)
Part 5 demonstrates vector addition in SYCL* C++ with a 'function' coding style using unified shared memory (USM). This code style will be familiar to users who are already experienced with SYCL*. Observe how the `VectorAdd` function is called using a lambda expression:
```c++
h.single_task<...>([=]() {
    VectorAdd(...);
});
```
```

I think this should be part 3. that way we only change one thing between part 2 and 3 (namely, we switch from a functor to a function).

Use the lambda explanation that we already have.

eventually, when mm host attributes can be applied to pointer arguments, we can stop advocating the functor model :)
